File: ESP8266/NodeOne_Python/NodeOne(Python)_V1.0/aREST.py
from machine import ADC,PWM
import json
from si7021 import *
class aREST:
	flag = 1
	def handle(self,client,request_handle):
		self.si7021 = si7021()
		pin = [16,5,4,0,2,14,12,13,15,3,1,9,10] #GPIO->pin
		mode = ''
		num = ''
		value = ''
		answer={}
		pin5 = Pin(pin[5],Pin.OUT)
		pin3 = Pin(pin[3],Pin.OUT)
		req = request_handle.decode('UTF-8','strict')
		c = client
		#find start
		e = req.find('/')
		req = req[e+1:]
		#find end
		e = req.find('HTTP')
		req = req[:e-1]
		#find mode
		e = req.find('/')
		if -1 == e:
			mode = req
		else:
			mode = req[:e]
			#find pin num
			req = req[e+1:]
			e = req.find('/')
			if -1 == e:
				num = req
			else:
				num = req[:e]
				#find value
				value = req[e+1:]
		if mode == 'exit':
			self.flag = 0
			answer['message'] = 'A socket to stop working,start again please restart'
		if mode == 'mode':
			if value == 'o':
				answer['message'] = 'Pin '+num+' set to output'
			if value == 'i':
				answer['message'] = 'Pin '+num+' set to input'
		if mode == 'digital':
			if value == '0':
				pin_out = Pin(pin[int(num)],Pin.OUT)
				pin_out.value(0)
				answer['message'] = 'Pin '+num+' set to 0'
			if value == '1':
				pin_out = Pin(pin[int(num)],Pin.OUT)
				pin_out.value(1)
				answer['message'] = 'Pin '+num+' set to 1'
			if value == 'r':
				pin_in = Pin(pin[int(num)],Pin.IN)
				read_value = pin_in.value()
				answer['message'] = 'Pin '+num+' read value '+str(read_value)
		if mode == 'analog':
			if num == '0':
				pin_out = Pin(pin[8],Pin.OUT)
				pin_out.value(1)
			if num == '1':
				pin_out = Pin(pin[8],Pin.OUT)
				pin_out.value(0)	
			adc_value = ADC(0).read()
			answer['message'] = 'A'+num+' value:'+str(adc_value)
		if mode == 'pwm':
			pin_pwm = PWM(Pin(pin[int(num)]),freq=100,duty=int(value))#pwm init
			answer['message'] = 'PWM:Pin '+num+' set to '+value
		if mode == 'temperature':
			answer['message'] = 'Temperature(¡æ):'+str(self.si7021.temp_value())
		if mode == 'humidity':
			answer['message'] = 'Humidity(%RH):'+str(self.si7021.humi_value())	
		if mode == 'forward':
			pin5.value(1)
			pin3.value(1)
			answer['message'] = 'car forward now...'	
		#send answer	
		c.send('HTTP/1.1 200 OK\r\nContent-type: text/html\r\nAccess-Control-Allow-Origin:* \r\n\r\n'+json.dumps(answer)+'\r\n')	
		
		

File: 2. Lua/arest.lua
--╪стьнд╪Ч temperature, PM2.5, RGB--
require "si7021"
dofile("PwmSvr_PCA9685.lua")
PwmSvr.begin()
PwmSvr.setPWMFreq(60)

local aREST = {}
local message = "Wrong API."

function aREST.handle(conn, request)

	local R_D0 = 0
	local R_D5 = 5
	local L_D4 = 4
	local L_D3 = 3
	
    -- New request, Find start/end
    local e = string.find(request, "/")
    local request_handle = string.sub(request, e + 1)
    e = string.find(request_handle, "HTTP")
    request_handle = string.sub(request_handle, 0, (e-2))
    print('-----------------------')
    print('Request: ', request_handle)
		
	-- if no favicon included, re-ew message everytime
	-- if favicon included, just use the last message
	if(string.find(request_handle,"favicon.ico") == nil) then message = "Wrong API." end 
	
	-- Pattern: http://IP/mode/value[2]/value[3]/value[4]
	local value={} ; i=1
	for str in string.gmatch(request_handle, "([^//]+)") do
		value[i] = str
		i = i + 1
	end
	local mode = value[1]
	if mode == nil then mode = "" end

	--------------Wifi---------------------
    if mode == "wifi"  and not check_nil(value, 3) then
        if value[2]~=nil and string.len(value[3])>=8 then
            file.open("config_wifi.lua","w+")
            value[2] = string.gsub(value[2],"+"," ")
            file.writeline('ssid="'..value[2]..'"')
            file.writeline('pwd="'..value[3]..'"')
            file.close()
            node.restart()
        end
	else
		value[2] = tonumber(value[2])
    end
    
    --------------General---------------------
    if mode == "mode" and not check_nil(value, 3) then
        if value[3] == "o" then
            gpio.mode(value[2], gpio.OUTPUT)
            message = "" .. value[2] .. " set to output" 
        elseif value[3] == "i" then
            gpio.mode(value[2], gpio.INPUT)
            message = "" .. value[2] .. " set to input"
        elseif value[3] == "p" then
            pwm.setup(value[2], 50, 0);
            pwm.start(value[2]);
            message = "Pin D" .. value[2] .. " set to PWM";
        end 
    end

    if mode == "digital" and not check_nil(value, 3) then
        if value[3] == "0" then 
            gpio.mode(value[2], gpio.OUTPUT)
            gpio.write(value[2], gpio.LOW)
            message = "" .. value[2] .. " set to 0"   
        elseif value[3] == "1" then
            gpio.mode(value[2], gpio.OUTPUT)
            gpio.write(value[2], gpio.HIGH)
            message = "" .. value[2] .. " set to 1" 
        elseif value[3] == "r" then
            value = gpio.read(value[2])
            message = value
        end
    end

    if (mode == "pwm" or mode == "output") and not check_nil(value, 3) then
		value[3] = error_handling(tonumber(value[3]),0,1023)
		pwm.setup(value[2],50,value[3])	
		pwm.start(value[2])
		message = ""..value[2]..":"..value[3]	
	end
    
    if (mode == "analog" or mode == "input") and not check_nil(value, 2) then
        gpio.mode(8,gpio.OUTPUT)
        if value[2] == 0 then
            gpio.write(8,gpio.HIGH)
        else
            gpio.write(8,gpio.LOW)
        end
        message = error_handling(adc.read(0),0,1023)
    end
	
    --------------Function port---------------------
    if (mode == "servo" or mode == "servo2") and not check_nil(value, 3) then
		value[3] = error_handling(tonumber(value[3]),0,180)
		if mode == "servo"  then
			pwm.setup(value[2],50,math.floor(33+((128-33)*value[3]/180)))
			pwm.start(value[2])
		elseif mode == "servo2" then			
			PwmSvr.setPWM(value[2], 0, math.floor(130+((550-130)*value[3]/180)))
		end
        message = ""..value[2]..":"..value[3]
    end
    
    if mode == "motor" and not check_nil(value, 4) then
		
		value[4] = error_handling(tonumber(value[4]),0,1023)
		
        if value[2] == 1 then 
            if value[3] == "cw" then 
				gpio.write(R_D0,gpio.HIGH)
                pwm.setduty(R_D5,value[4])
                message = "Motor " .. value[2] .. " set to " .. value[4] .. " in " .. value[3]   
            elseif value[3] == "acw" then
				gpio.write(R_D0,gpio.LOW)
                pwm.setduty(R_D5,value[4])
                message = "Motor " .. value[2] .. " set to " .. value[4] .. " in " .. value[3]  
            end 
        elseif value[2] == 2 then
            if value[3] == "cw" then 
				gpio.write(L_D4,gpio.HIGH)
                pwm.setduty(L_D3,value[4])
                message = "Motor " .. value[2] .. " set to " .. value[4] .. " in " .. value[3]   
            elseif value[3] == "acw" then
				gpio.write(L_D4,gpio.LOW)
                pwm.setduty(L_D3,value[4])
                message = "Motor " .. value[2] .. " set to " .. value[4] .. " in " .. value[3]  
            end 
        end
    end
    
	if(value[2] ~= nil) then value[2] = error_handling(tonumber(value[2]),0,1023) end
	if mode == "forward" and not check_nil(value, 2) then 
		message = motor_control(value[2],value[2],0,1,"forward",value[2])
	elseif mode == "backward" and not check_nil(value, 2) then
		message = motor_control(value[2],value[2],1,0,"backward",value[2])
	elseif  mode == "left" and not check_nil(value, 2) then
		message = motor_control(value[2],value[2],1,1,"left",value[2])
	elseif mode == "right" and not check_nil(value, 2) then
		message = motor_control(value[2],value[2],0,0,"right",value[2])
	elseif mode == "stop" and not check_nil(value, 1) then
		message = motor_control(0,0,0,0,"stop","")
    end	
                   
    if mode == "temperature" then
        local temp = read_temp()
        message = ""..temp	
    end
    
    if mode == "humidity" then
        local humi = read_humi()
        message = ""..humi	
    end
        
    conn:send("HTTP/1.1 200 OK\r\nContent-type: text/html\r\nAccess-Control-Allow-Origin:* \r\n\r\n" .. message .. "\r\n")
end

function motor_control(M2_speed,M1_speed,M2_direction,M1_direction,direction,speed)
	pwm.setduty(5,M2_speed)
	pwm.setduty(3,M1_speed) 
	gpio.write(0,M2_direction)
	gpio.write(4,M1_direction)
	return "car "..direction.." "..speed.." now... "
	
end

function error_handling(value,min_value,max_value)
	num = value
	if num <= min_value then
		num = min_value
	elseif  num >= max_value then
		num = max_value
	end
	return num
end

function check_nil(value, num)
	for i=1,num,1 do
		if value[i] == nil then
			return true
		end
	end
	return false
end

return aREST